soc/nordic/nrf54h/poweroff.c

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/sys/poweroff.h>
+#include <zephyr/toolchain.h>
+#include <zephyr/pm/policy.h>
+#include <zephyr/arch/common/pm_s2ram.h>
+#include <hal/nrf_resetinfo.h>
+#include <hal/nrf_lrcconf.h>
+#include <hal/nrf_memconf.h>
+#include <zephyr/cache.h>
+
+#if defined(NRF_APPLICATION)
+#define RAMBLOCK_CONTROL_BIT_ICACHE 1
+#define RAMBLOCK_CONTROL_BIT_DCACHE 2
+#define RAMBLOCK_POWER_ID	    0
+#define RAMBLOCK_CONTROL_OFF	    0
+#elif defined(NRF_RADIOCORE)
+#define RAMBLOCK_CONTROL_BIT_ICACHE 2
+#define RAMBLOCK_CONTROL_BIT_DCACHE 3
+#define RAMBLOCK_POWER_ID	    0
+#define RAMBLOCK_CONTROL_OFF	    0
+#else
+#error "Unsupported domain."
+#endif
+
+static void suspend_common(void)
+{
+	if (IS_ENABLED(CONFIG_DCACHE)) {
+		/* Flush, disable and power down DCACHE */
+		sys_cache_data_flush_all();
+		sys_cache_data_disable();
+		nrf_memconf_ramblock_control_enable_set(NRF_MEMCONF, RAMBLOCK_POWER_ID,
```

Comment:
A little updated procedure is described in this PR: https://github.com/nrfconnect/doc-internal/pull/93